internal/mesh/internal/controllers/sidecarproxy/mapper.go

```diff
@@ -0,0 +1,217 @@
+		return nil, err
+	}
+
+	return controller.MakeRequestsFromResources(pbmesh.ProxyStateTemplateType, workloads), nil
+}
+
+func MapComputedTrafficPermissions(ctx context.Context, rt controller.Runtime, res *pbresource.Resource) ([]controller.Request, error) {
+	assertResourceType(pbauth.ComputedTrafficPermissionsType, res.Id.Type)
+
+	wiID := resource.ReplaceType(pbauth.WorkloadIdentityType, res.Id)
+
+	workloads, err := rt.Cache.List(pbcatalog.WorkloadType, workloadByWorkloadIdentityIndex.Name(), wiID)
+	if err != nil {
+		return nil, err
+	}
+
+	return controller.MakeRequestsFromResources(pbmesh.ProxyStateTemplateType, workloads), nil
+}
+
+func MapComputedRoutes(ctx context.Context, rt controller.Runtime, res *pbresource.Resource) ([]controller.Request, error) {
+	assertResourceType(pbmesh.ComputedRoutesType, res.Id.Type)
+
+	svcID := resource.ReplaceType(pbcatalog.ServiceType, res.Id)
+
+	return mapServiceThroughDestinations(rt.Cache, svcID)
+}
+
+func MapService(ctx context.Context, rt controller.Runtime, res *pbresource.Resource) ([]controller.Request, error) {
+	assertResourceType(pbcatalog.ServiceType, res.Id.Type)
+
+	pstIDs, err := mapServiceThroughDestinations(rt.Cache, res.Id)
+	if err != nil {
+		return nil, err
+	}
+
+	// Now walk the mesh configuration information backwards because
+	// we need to find any PST that needs to DISCOVER endpoints for this
+	// service as a part of mesh configuration and traffic routing.
+
+	// Find all ComputedRoutes that reference this service.
+	computedRoutes, err := rt.Cache.List(
+		pbmesh.ComputedRoutesType,
+		computedRoutesByBackendServiceIndex.Name(),
+		res.Id,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, cr := range computedRoutes {
+		// Find all Upstreams that reference a Service aligned with this
+		// ComputedRoutes. Afterwards, find all Workloads selected by the
+		// Upstreams, and align a PST with those.
+
+		ids, err := MapComputedRoutes(ctx, rt, cr)
+		if err != nil {
+			return nil, err
+		}
+		pstIDs = append(pstIDs, ids...)
+	}
+
+	return pstIDs, nil
+}
+
+func mapServiceThroughDestinations(cache cache.ReadOnlyCache, svcID *pbresource.ID) ([]controller.Request, error) {
+	explicitDests, err := cache.List(
+		pbmesh.ComputedExplicitDestinationsType,
+		computedExplicitDestinationsByServiceIndex.Name(),
+		svcID,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	pstMap := make(map[resource.ReferenceKey]*pbresource.ID)
+	for _, dest := range explicitDests {
+		pstID := resource.ReplaceType(pbmesh.ProxyStateTemplateType, dest.Id)
+		rk := resource.NewReferenceKey(pstID)
+		if _, ok := pstMap[rk]; !ok {
+			pstMap[rk] = pstID
+		}
+	}
+
+	implicitDests, err := cache.List(
+		pbmesh.ComputedImplicitDestinationsType,
+		computedImplicitDestinationsByServiceIndex.Name(),
+		svcID,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	reqs := make([]controller.Request, 0, len(explicitDests)+len(implicitDests))
+	reqs = append(reqs, controller.MakeRequestsFromResources(
```

Comment:
It might be worth commenting that explicit destinations are name aligned with PSTs. The difference between implicit and explicit destinations threw me off for a few minutes.